handlers/captcha.ts

```diff
@@ -0,0 +1,168 @@
+        .callback_data.split(":");
+      slices[1] = slices[1] == emoji ? replacement : slices[1];
+      return {
+        ...v,
+        text: slices[1] == emoji ? replacement : slices[1],
+        callback_data: slices.join(":"),
+      };
+    })
+  );
+
+composer.callbackQuery(/^emoji-captcha:([^:]+):/, async (ctx) => {
+  await ctx.answerCallbackQuery();
+  const emoji = ctx.match![0];
+  if ([EMOJI_CORRECT, EMOJI_WRONG].includes(emoji)) {
+    return;
+  }
+  const chatId = Number(
+    (ctx.msg?.reply_to_message?.reply_markup
+      ?.inline_keyboard[0][0] as InlineKeyboardButton.CallbackButton)
+      .callback_data,
+  );
+  const buttons = ctx.msg?.reply_markup
+    ?.inline_keyboard as InlineKeyboardButton.CallbackButton[][];
+  const correctEmojis = (await decrypt(
+    buttons.flat().map((v) => v.callback_data.split(":")[2] ?? "").join(""),
+  ))
+    .split(";");
+  const attempts = buttons.flat().filter((v) =>
+    [EMOJI_CORRECT, EMOJI_WRONG].includes(v.text)
+  ).length + 1;
+  const previousWrongAttempts = buttons.flat().filter((v) =>
+    v.text == EMOJI_WRONG
+  ).length;
+  if (emoji == EMOJI_WRONG || emoji == EMOJI_CORRECT) {
+    return;
+  }
+  if (correctEmojis.includes(emoji)) {
+    await ctx.editMessageReplyMarkup({
+      reply_markup: {
+        inline_keyboard: replaceEmoji(buttons, emoji, EMOJI_CORRECT),
+      },
+    });
+    if (attempts == 6) {
+      await ctx.deleteMessage();
+      await ctx.api.editMessageText(
+        ctx.chat?.id!,
+        ctx.msg?.reply_to_message?.message_id!,
+        "You're now in!",
+      );
+      await ctx.api.approveChatJoinRequest(chatId, ctx.chat?.id!);
+    }
+  } else {
+    await ctx.editMessageReplyMarkup({
+      reply_markup: {
+        inline_keyboard: replaceEmoji(buttons, emoji, EMOJI_WRONG),
+      },
+    });
+    if (previousWrongAttempts == 2) {
+      await ctx.deleteMessage();
+      await ctx.api.editMessageText(
+        ctx.chat?.id!,
+        ctx.msg?.reply_to_message?.message_id!,
+        "You couldn't make it. Try again later.",
+      );
+    }
+  }
+});
+
+async function emojiCaptcha(
+  ctx: CallbackQueryContext<Context>,
+) {
+  const res = await fetch("REPLACE_WITH_ENV_VAR");
+  const emojis = getEmojis(res.headers.get("x-emojis") ?? "");
+  const correctEmojis = getEmojis(res.headers.get("x-correct-emojis") ?? "");
+  if (emojis.length == 0 || correctEmojis.length == 0) {
+    return;
+  }
+  let encrypted = await encrypt(correctEmojis.join(";"));
+  const keyboard = new InlineKeyboard();
+  for (let i = 0; i < emojis.length; i += BUTTONS_PER_ROW) {
+    for (
+      const emoji of emojis.slice(i, i + BUTTONS_PER_ROW)
+    ) {
+      let data = `emoji-captcha:${emoji}:`;
+      const available = 64 - new TextEncoder().encode(data).length;
+      data += encrypted.slice(0, available);
+      encrypted = encrypted.slice(available);
+      keyboard.text(emoji, data);
+    }
+    keyboard.row();
+  }
```

Comment:
A suggestion (Personally I think its better than this slicing and nested loops).

https://github.com/dcdunkan/ryportbot/blob/main/bot.ts#L306-L310

Reply:
I don't know how that should be implemented here.